plugin/evm/limit_order.go

```diff
@@ -47,38 +54,55 @@ func (lop *limitOrderProcesser) ListenAndProcessTransactions() {
 	lastAccepted := lop.blockChain.LastAcceptedBlock().NumberU64()
 	if lastAccepted > 0 {
 		log.Info("ListenAndProcessTransactions - beginning sync", " till block number", lastAccepted)
-
-		allTxs := types.Transactions{}
-		for i := uint64(0); i <= lastAccepted; i++ {
-			block := lop.blockChain.GetBlockByNumber(i)
-			if block != nil {
-				for _, tx := range block.Transactions() {
-					if lop.limitOrderTxProcessor.CheckIfOrderBookContractCall(tx) {
-						lop.limitOrderTxProcessor.HandleOrderBookTx(tx, i, *lop.backend)
-					}
-				}
-			}
+		ctx := context.Background()
+
+		filterSystem := filters.NewFilterSystem(lop.backend, filters.Config{})
+		filterAPI := filters.NewFilterAPI(filterSystem, true)
+		logs, err := filterAPI.GetLogs(ctx, filters.FilterCriteria{
+			Addresses: []common.Address{limitorders.OrderBookContractAddress, limitorders.ClearingHouseContractAddress, limitorders.MarginAccountContractAddress},
+		})
+		if err != nil {
+			log.Error("ListenAndProcessTransactions - GetLogs failed", "err", err)
+			panic(err)
 		}
-
-		log.Info("ListenAndProcessTransactions - sync complete", "till block number", lastAccepted, "total transactions", len(allTxs))
+		log.Info("ListenAndProcessTransactions", "total logs", len(logs), "err", err)
+		processEvents(logs, lop)
 	}
 
 	lop.listenAndStoreLimitOrderTransactions()
 }
 
-func (lop *limitOrderProcesser) RunMatchingEngine() {
+func (lop *limitOrderProcesser) IsFundingPaymentTime(lastBlockTime uint64) bool {
+	return lastBlockTime >= lop.memoryDb.GetNextFundingTime()
+}
+
+func (lop *limitOrderProcesser) ExecuteFundingPayment() error {
+	// @todo get index twap for each market with warp msging
+
+	return lop.limitOrderTxProcessor.ExecuteFundingPaymentTx()
+}
+
+func (lop *limitOrderProcesser) RunLiquidationsAndMatching() {
 	lop.limitOrderTxProcessor.PurgeLocalTx()
-	longOrders := lop.memoryDb.GetLongOrders()
-	shortOrders := lop.memoryDb.GetShortOrders()
+	for _, market := range limitorders.GetActiveMarkets() {
+		longOrders := lop.memoryDb.GetLongOrders(market)
+		shortOrders := lop.memoryDb.GetShortOrders(market)
+		longOrders, shortOrders = lop.runLiquidations(market, longOrders, shortOrders)
```

Comment:
I would prefer not mutating arguments especially in stuff we do during build block. 
It will keeps the pipeline(liquidation -> matching -> whatever else we add in future) clean and allows us to design it as more as functional code.
Also the code is easier to understand(less chance of being lost in mutation) and test. 
I don't think we will have major memory issues but if we start encountering those we can then make these changes.

Reply:
though here we are mutating the slice (longOrders and shortOrders) but we are replacing its contents. So its not all immutable.

Reply:
Understood. Thank you